app/gui2/shared/yjsModel.ts

```diff
@@ -70,70 +70,81 @@ export class DistributedProject {
   }
 }
 
-export class DistributedModule {
-  doc: Y.Doc
+export class ModuleDoc {
+  ydoc: Y.Doc
   contents: Y.Text
   idMap: Y.Map<Uint8Array>
   metadata: Y.Map<NodeMetadata>
+  constructor(ydoc: Y.Doc) {
+    this.ydoc = ydoc
+    this.contents = ydoc.getText('contents')
+    this.idMap = ydoc.getMap('idMap')
+    this.metadata = ydoc.getMap('metadata')
+  }
+}
 
-  static async load(doc: Y.Doc) {
-    doc.load()
-    await doc.whenLoaded
-    return new DistributedModule(doc)
+export class DistributedModule {
+  doc: ModuleDoc
+
+  static async load(ydoc: Y.Doc) {
```

Comment:
I'm not sure. In Rust, you *have to* specify return types on each function. That's fine, because the language is so strongly typed in the first place that you almost always have the type nicely defined somewhere, so it's easy to just write it there. In typescript, it's often the case that the return type is a complex object of an anonymous type. Then you speciifcally want to avoid specifying return type to avoid repeating yourself (e.g. in store definitions).

I'm fine with *preferring* to specify return types in anything that's public/exported, especially when the type is already easy to specify. But I'm not sure if encoding the exact scenarios where it should be *required* to a lint is a net positive. Too strict linting rules get annoying really fast, and there are always exceptions. I really don't like having to specify `eslint-ignore-next-line` everywhere.

Reply:
i'm not sure it's different in rust though - rather it's that you *can't* return anonymous objects, so you *have to* have the struct already defined to be able to create the object in the first place.
but yeah, definitely agree that store definitions are one (of the few, imo) exceptions